sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java

```diff
@@ -0,0 +1,98 @@
+ * However, watermark is never set in future and capped to 'now - max delay'. In addition, watermark
+ * advanced to 'now - max delay' when the queue has caught up (previous read attempt returned no
+ * message)
+ *
+ * @see "org.apache.beam.sdk.io.kafka.CustomTimestampPolicyWithLimitedDelay for the inspiration and
+ *     corresponding KafkaIO policy"
+ */
+public class CustomTimestampPolicyWithLimitedDelay extends TimestampPolicy {
+  private final Duration maxDelay;
+  private final SerializableFunction<RabbitMqMessage, Instant> timestampFunction;
+  private Instant maxEventTimestamp;
+
+  /**
+   * A policy for custom record timestamps where timestamps are expected to be roughly monotonically
+   * increasing with out of order event delays less than {@code maxDelay}. The watermark at any time
+   * is {@code Min(now(), max_event_timestamp) - maxDelay}.
+   *
+   * @param timestampFunction A function to extract timestamp from the record
+   * @param maxDelay For any record in the Kafka partition, the timestamp of any subsequent record
+   *     is expected to be after {@code current record timestamp - maxDelay}.
+   * @param previousWatermark Latest check-pointed watermark
+   */
+  public CustomTimestampPolicyWithLimitedDelay(
+      Duration maxDelay,
+      SerializableFunction<RabbitMqMessage, Instant> timestampFunction,
+      Optional<Instant> previousWatermark) {
+    this.maxDelay = maxDelay;
+    this.timestampFunction = timestampFunction;
+    // 'previousWatermark' is not the same as maxEventTimestamp (e.g. it could have been in future).
+    // Initialize it such that watermark before reading any event same as previousWatermark.
+    maxEventTimestamp = previousWatermark.orElse(BoundedWindow.TIMESTAMP_MIN_VALUE).plus(maxDelay);
+  }
+
+  @Override
+  public Instant getTimestampForRecord(RabbitMqMessage record) {
+    Instant ts = timestampFunction.apply(record);
+    if (ts.isAfter(maxEventTimestamp)) {
+      maxEventTimestamp = ts;
+    }
+    return ts;
+  }
+
+  @Override
+  public Instant getWatermark(LastRead ctx, RabbitMqMessage record) {
+    // Watermark == maxEventTime - maxDelay, except in two special cases:
+    //   a) maxEventTime in future : probably due to incorrect timestamps. Cap it to 'now'.
+    //   b) queue is empty : Need to advance watermark if there are no records in the queue.
+    //         We assume that future records will have timestamp >= 'now - maxDelay' and advance
+    //         the watermark accordingly.
+    // The above handles majority of common use cases for custom timestamps. Users can implement
+    // their own policy if this does not work.
+
+    return getWatermark(ctx, Instant.now());
+  }
+
+  @VisibleForTesting
+  Instant getWatermark(LastRead ctx, Instant now) {
+    if (maxEventTimestamp.isAfter(now)) {
+      return now.minus(maxDelay); // (a) above.
+    } else if (ctx.getMessageBacklog() == 0
```

Comment:
Now that we have accurate (I hope?) message backlog values, this now behaves about exactly as Kafka's does